week-2/extended-project/yoda-y-sus-apis/web_api.rb

```diff
@@ -0,0 +1,208 @@
+# characters
+#
+# name    : string
+# films   : array
+# gender  : string
+# species : array
+# id      : string
+
+characters_modified = characters.map do |value|
+  {
+    name: value["name"],
+    films: value["films"],
+    gender: value["gender"],
+    species: value["species"],
+    id: value["url"]
+  }
+end
+
+# films
+#
+# episode_id : integer
+# title      : string
+# director   : string
+# release    : string
+# characters : array
+# id         : string
+#
+
+films_modified = films.map do |value|
+  {
+    episode_id: value["episode_id"],
+    title: value["title"],
+    director: value["director"],
+    release: value["release_date"],
+    characters: value["characters"],
+    id: value["url"]
+  }
+end
+
+# planets
+#
+# name      : string
+# residents : array
+# id        : string
+
+planets_modified = planets.map do |value|
+  {
+    name: value["name"],
+    residents: value["residents"],
+    id: value["url"]
+  }
+end
+
+# species
+# name : string
+# id   : string
+
+species_modified = species.map do |value|
+  {
+    name: value["name"],
+    id: value["url"]
+  }
+end
+
+#Convert data to information
+#=====================================================
+# Attach all the string content about SW characters
+# Required content: films played by each character
+#										specie which each character belongs
+
+character_final = characters_modified.map do |value|
+	films_character = value[:films]
+	# catching film's titles and number of episodes
+  new_value = films_character.map do |film_id|
+									index = films_modified.index { |film| film[:id] == film_id }
+									film_match = films_modified[index]
+                  "Episode " + film_match[:episode_id].to_s + ": " + film_match[:title]
+              end
+	# catching specie name
+  index_species = species_modified.index { |specie| specie[:id] == value[:species][0] }
+  
+	# delivering data
+  value[:films] = new_value.sort.join("|")
+  value[:species] = index_species != nil ? species_modified[index_species][:name] : "Without specie"
+  value
+end
+
+# Attach all the string content about SW films
+# Required content: characters that appear in each film
+#										films' titles
+
+films_final = films_modified.map do |value|
+  character_film = value[:characters]
+  new_value=[]
+# catch characters names in planet
+  new_value[0] = character_film.map do |character_id|
+    index = characters_modified.index { |character| character[:id] == character_id }
+    characters_modified[index][:name]
+  end
+# catching characters genders in planet
+  new_value[1] = character_film.map do |character_id|
+    index = characters_modified.index { |character| character[:id] == character_id }
+    characters_modified[index][:gender]
+	end
+# catching score of film
+	score = get_score(value[:title])
+# delivering data
+	value[:characters] = new_value[0].sort.join("|")
+  value[:score] = score
+  value[:gender_count] = new_value[1].group_by {|e| e}.map{|k, v| "#{k} - #{v.length}"}.join("|")
```

Comment:
I don't understand this line. Could you explain me how works?

Reply:
You probably want to split that line to make it more understandable!

Also, if you're dealing with maps inside maps, it's best to start thinking in what data are you extracting, and create methods for each step. Helps a lot when dealing with complex logic!

Reply:
value[:gender_count] = new_value[1].group_by {|e| e}
                                                          .map{|k, v| "#{k} - #{v.length}"}
                                                          .join("|")

The idea here is: 
- group the results provided by the API
- count the elements in the returned array
- join the array in a string like: Droid - 1\|Human - 5\|Gungan - 3\|Without specie - 2